common/ipc.ts

```diff
@@ -23,6 +23,11 @@ const ipcRenderer =
     : null
 const ipcMain = IS_ELECTRON && !IS_RENDERER ? require('electron').ipcMain : null
 
+export interface FsResult<T> {
```

Comment:
I'm not a big fan of defining types in this way, where you have multiple fields which are exclusionary with one another. These functions that we're typing here return an error with an error code _OR_ they return a payload for that function. Mixing those fields together in one type usually means you have to make some or all of them optional, which makes the types less correct. If the `fsStat` method returns successfully it needs to have a certain payload that is not optional. Another benefit would be not having this ambiguous property `payload`, and instead have a proper name for field(s) each function returns.

I would prefer to have something similar to what I showed on the Discord, or to at least try to make it work with what tec suggested, by using the user defined type guards that you linked from TS documentation. For what it's worth, I think we're using the former pattern more frequently throughout the code.

Reply:
The issue I ran into with the solutions we were talking about in the discord was that in the case of `fsReadDir`, and `fsReadFile`, I couldn't add an `isError` field without creating a new result type to wrap `ArrayBuffer` and `FsDirent[]`. I figured since I needed to wrap those types, I could consolidate all the calls into a consistent API.

What do you think about this https://github.com/ShieldBattery/ShieldBattery/pull/939/commits/16b68b58710aeb7d0bd78df01c9dac882d943991? By using the type union, we know at compile time if the result has an error or has a value. It'll still have the issue of an ambiguous field name, but with context and type information it should be pretty clear most the time

I still need to cleanup some stuff, but I wanted to give you a chance to give the general API 👍/👎

Reply:
>The issue I ran into with the solutions we were talking about in the discord was that in the case of fsReadDir, and fsReadFile, I couldn't add an isError field without creating a new result type to wrap ArrayBuffer and FsDirent[].

I don't think this is really an issue at all. In fact, I would consider this a benefit since the type is very simple, you immediately know what it contains from its property names, and it's easy to extend in the future (in case we need to add something to it). Also, if you make the `error` (or `isError`) property optional like tec suggested in the Discord, it becomes even simpler to use.

Trying to consolidate types here seem like a more complex solution to me with very little benefit.

I've updated my [TypeScript playground](https://www.typescriptlang.org/play?#code/KYOwrgtgBAYgzgUQE5IPZIMKoCbCgbwFgAoKKAWQEs45KQBzGSgGzwF4oByKmuxl4JwA0JMj1oMYqZriRQO3ahMbTZwkgF8SJOgBdgSAGYBDAMZ4AQsbjB4AJWBwwzXQVFQDaJAH4AXFAAjVGlgYxBNbWI9AxNzWEQUdA8AD30QbDgoKxt7R2dXIlIPRKR-XSQwYHdTHGB-eGQvLFwI4h0QfSMzPFzjbAARSjlgVNAMrOtbOAcnFzciz3Q-KBNmG3dDATh-OHK+AG0AXVb2ztie6dDsJlYUtPHsqZn8+bJFn39V9aLN1n8AQRQxgAnhYwIZDAYTsRDGAQKZdJRUCAVpc+oMkAAKACU9TRAyGUAAPvFGklCmRKIYoJiAIzY15kKBIYC6MBIFEUplkX6OfxHERFMhaIoaDxrPBcsgstkcxnc95lCrAQXcsg1XB4smYWoAOnEfCkMgMqqZIuF0Nh8MRyNRDj6N2AOLx9uuAmJpJKjKpNPp8uZrPZnPc3N5-hAwAA7lBAUgQWCIQY6djTRbReKbP6ZUH-W8SkrKqmmRq6p6mnqDZIBEXzVARSKSDUQLsA7NdLT5HarhicZEfZiWW3abr3gyuQB6cet-Id6hlpIACwMVSKTbgIV1zFQ9AHeRcw5L2M0Gcl7kn0-3UDnvQJciXLOqyPXrE3293Q91vLgR+IDeIa9cQd8gAJk7Qx8UdXs2iial3xAkcSjHM8pyAlxQOvBIvCge8V3VJ8Ny3HdUN0YDdUPY9gAlRlz2I9DMhvR1sOXR9mwIt9aM-AQfxFIA) to show how it would look like for both the `fsReadDir` and `fsReadFile`.

Reply:
Fair enough. I updated let me know if it's good and I"ll squash the commits.

A small annoyance is that it looks like Promise is confusing the structural typing with the type union so I have to manually set them to variables before returning. He's a [playground ](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAMwM4AUoRMJBTJj0hwDkyA9GAVrkQLABQ92AdgK6EBiSAolOlAMIQAJtjgBvenDgBZLEmBMA5u2AAbUQF5ispPKUr1RADSSZchcoiqRUOFqI69l69ijH6AX3r0FMVwgBDAGNRACEAnE4AJVwWVXgJOilXPgB+AC44ACMIK2wApk9vOl9-YNFOHj44bAAPPyYhPHDIpBikOITTFOhMmCgWbFMg4WxMyt5oQREihhKmPyhAkLho-KEAEWBbOoamuBbsNY748W7JqAzEANUcUwQ1XEykfosAbQBdWZ8FspW1gJCAyiXbMfaHY6dM5JGoXK6BW5DGEPdSZACCvACAE9QiwEAhXN86AgWEwgjBgBAmIg2ustlAABQASkyaAwVAAPADNts4AAfVbcC4APmhUmACDgDIAjEyxVI4FBsDAWFBqYkFQqUU84J8TDCpF4YR4aoj5VIlSq1eaFT0oH0Bth9ZqpCMROMhXxptgAHSOCzsKw2Z0Ko2GokkskUqk0mKA4HM1mUHBc2nxx6igVszApiZ8UUauASqWym2W1Xq0ya7WZJjYADucAxUGxuPxrhlTJD4eNppwZeVFZtyQuDsG3YVbrGgqqU1GfvM+keE7DcCNRvoIyYL0VsXi0rssbp22ZxWLDKVJxg0p9drlhdIpF3V4PWBnFzgAAtXEjXVSkHkPqqBAigXnu14+lOTKeH2ogPk+l6dK+eDcvSX4-sM-6AcBoGIfuPrakg0F0BudBbjueEwAATIeyBxkCjynnM56UVRt4XPepiPs+nQ0W+ebQOhSqYdu2EgWBV5sVBMHYGa8E8fEfEoWmDHqEJv5wORYm4eBbHasRRpAA) illustrating it so you can see the error

Reply:
Without manually specifying error on the result types, I'd need to add a https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards for each I think, or have to cast

Reply:
>A small annoyance is that it looks like Promise is confusing the structural typing with the type union so I have to manually set them to variables before returning. He's a [playground ](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAMwM4AUoRMJBTJj0hwDkyA9GAVrkQLABQ92AdgK6EBiSAolOlAMIQAJtjgBvenDgBZLEmBMA5u2AAbUQF5ispPKUr1RADSSZchcoiqRUOFqI69l69ijH6AX3r0FMVwgBDAGNRACEAnE4AJVwWVXgJOilXPgB+AC44ACMIK2wApk9vOl9-YNFOHj44bAAPPyYhPHDIpBikOITTFOhMmCgWbFMg4WxMyt5oQREihhKmPyhAkLho-KEAEWBbOoamuBbsNY748W7JqAzEANUcUwQ1XEykfosAbQBdWZ8FspW1gJCAyiXbMfaHY6dM5JGoXK6BW5DGEPdSZACCvACAE9QiwEAhXN86AgWEwgjBgBAmIg2ustlAABQASkyaAwVAAPADNts4AAfVbcC4APmhUmACDgDIAjEyxVI4FBsDAWFBqYkFQqUU84J8TDCpF4YR4aoj5VIlSq1eaFT0oH0Bth9ZqpCMROMhXxptgAHSOCzsKw2Z0Ko2GokkskUqk0mKA4HM1mUHBc2nxx6igVszApiZ8UUauASqWym2W1Xq0ya7WZJjYADucAxUGxuPxrhlTJD4eNppwZeVFZtyQuDsG3YVbrGgqqU1GfvM+keE7DcCNRvoIyYL0VsXi0rssbp22ZxWLDKVJxg0p9drlhdIpF3V4PWBnFzgAAtXEjXVSkHkPqqBAigXnu14+lOTKeH2ogPk+l6dK+eDcvSX4-sM-6AcBoGIfuPrakg0F0BudBbjueEwAATIeyBxkCjynnM56UVRt4XPepiPs+nQ0W+ebQOhSqYdu2EgWBV5sVBMHYGa8E8fEfEoWmDHqEJv5wORYm4eBbHasRRpAA)illustrating it so you can see the error

I think the problem here is returning a Promise from a non-async function. If you make that function async, the error should go away, e.g. [TS playground](https://www.typescriptlang.org/play?#code/KYOwrgtgBAYgzgUQE5IPZIMKoCbCgbwFgAoKKAWQEs45KQBzGSgGzwF4oByKmuxl4JwA0JMj1oMYqZriRQO3ahMbTZwkgF8SJOgBdgSAGYBDAMZ4AQsbjB4AJWBwwzXQVFQDaJAH4AXFAAjVGlgYxBNbWI9AxNzWEQUdA8AD30QbDgoKxt7R2dXIlIPRKR-XSQwYHdTHGB-eGQvLFwI4h0QfSMzPFzjbAARSjlgVNAMrOtbOAcnFzciz3Q-KBNmG3dDATh-OHK+AG0AXVb2ztie6dDsJlYUtPHsqZn8+bJFn39V9aLN1n8AQRQxgAnhYwIZDAYTsRrMCQKYVmB4bpKKgQCtLn1BkgABQASn8AAU0BBqMAADy9AZDKAAH3ijXQAD5XlBKIYoDiAIx41lkJDAXRgJDowpkcUrLb+I4iIpkLRFDQeNZ4MXigVCkV88XvMoVYCyiVkGq4eoJJq1AB04j4UhkBkN4oV8uhsPhiORqPRhkx1wE+KJJLJlN9Nzw9IaJRZavZnJ52o1wtF7glvzqUBAwAA7lBAUgQWCIQZuXjHS7FcqbAnBUntW8SnrKmXxSb05GLbhrUpbQJm86oAqFSQaiBdlABbNdFz5FBjFnjJRXD6HFihvjIrGcRP8lzLe9eWqAPSH8d5FzT6gMkpQAAWBiqRRHcBCluYqHoW7PU8trbxmkrqruMep6ThemRUtit73tUaLPqwr7vp+oGWmmcB-sQQ7EE+rjbi4ABMM5zguS6hv66E6BySH5Hhe4lAeQEnrhugEZe7ZJHeAowaOL5vh+TE0b+-7ACqrLAfxbLgaRtwcQ+xqwTxiH8ShAjoZhIwAA7oAUCpAA)

Reply:
whoops, i should have looked at the playground error more closely. This is the error in the actual code. It looks like it's trying to combine the two return objects when building the structural type information? 

```
TS2345: Argument of type '(_: IpcMainInvokeEvent, dirPath: string, options: { withFileTypes: true; }) => Promise<{ files: FsDirent[]; error: boolean; code?: undefined; } | { code: FsErrorCode; error: boolean; files?: undefined; }>' is not assignable to parameter of type '(event: IpcMainInvokeEvent, dirPath: string, options: { withFileTypes: true; }) => Promisable<Promise<FsDirents> | Promise<FsError>>'.   Type 'Promise<{ files: FsDirent[]; error: boolean; code?: undefined; } | { code: FsErrorCode; error: boolean; files?: undefined; }>' is not assignable to type 'Promisable<Promise<FsDirents> | Promise<FsError>>'.     Type 'Promise<{ files: FsDirent[]; error: boolean; code?: undefined; } | { code: FsErrorCode; error: boolean; files?: undefined; }>' is not assignable to type 'Promise<FsDirents>'.       Type '{ files: FsDirent[]; error: boolean; code?: undefined; } | { code: FsErrorCode; error: boolean; files?: undefined; }' is not assignable to type 'FsDirents'.         Type '{ files: FsDirent[]; error: boolean; code?: undefined; }' is not assignable to type 'FsDirents'.           Types of property 'error' are incompatible.             Type 'boolean' is not assignable to type 'false'.
```

Reply:
I'm not sure where exactly do you get this error, but maybe it will go away once you address the other comments I left now 🤞

Reply:
Full Error:

```
TS2345: Argument of type '(_: IpcMainInvokeEvent, dirPath: string, options: { withFileTypes: true; }) => Promise<{ entries: FsDirent[]; code?: undefined; error?: undefined; } | { code: FsErrorCode; error: boolean; entries?: undefined; }>' is not assignable to parameter of type '(event: IpcMainInvokeEvent, dirPath: string, options: { withFileTypes: true; }) => Promisable<Promise<FsReadDirResult | FsError>>'.   Type 'Promise<{ entries: FsDirent[]; code?: undefined; error?: undefined; } | { code: FsErrorCode; error: boolean; entries?: undefined; }>' is not assignable to type 'Promisable<Promise<FsReadDirResult | FsError>>'.     Type 'Promise<{ entries: FsDirent[]; code?: undefined; error?: undefined; } | { code: FsErrorCode; error: boolean; entries?: undefined; }>' is not assignable to type 'Promise<FsReadDirResult | FsError>'.       Type '{ entries: FsDirent[]; code?: undefined; error?: undefined; } | { code: FsErrorCode; error: boolean; entries?: undefined; }' is not assignable to type 'FsReadDirResult | FsError'.         Type '{ code: FsErrorCode; error: boolean; entries?: undefined; }' is not assignable to type 'FsReadDirResult | FsError'.           Type '{ code: FsErrorCode; error: boolean; entries?: undefined; }' is not assignable to type 'FsError'.             Types of property 'error' are incompatible.               Type 'boolean' is not assignable to type 'true'.
``` 

If i change `error` to boolean for both the result types and error i'm able to return the object without the variable, but that breaks type inference when it's called and we'd need to cast. I can also return one without a variable but not the other without the type checker complaining 😅 .

I can also hoist the (lambda?) into a helper function and do something like 

```
  async function fsStat(filePath: string): Promise<FsStatResult | FsError> {
    try {
      const result = await fsPromises.stat(filePath)
      return {
        stats: {
          isFile: result.isFile(),
          isDirectory: result.isDirectory(),
          size: result.size,
          blksize: result.blksize,
          blocks: result.blocks,
          atimeMs: result.atimeMs,
          mtimeMs: result.mtimeMs,
          ctimeMs: result.ctimeMs,
          birthtimeMs: result.birthtimeMs,
          atime: result.atime,
          mtime: result.mtime,
          ctime: result.ctime,
          birthtime: result.birthtime,
        },
      }
    } catch (e) {
      if ((e as any).code === 'ENOENT') {
        return {
          error: true,
          code: FsErrorCode.FileOrFolderMissing,
        }
      }
      throw e
    }
  }
  ipcMain.handle('fsStat', async (_, filePath) => fsStat(filePath))
  ```

Reply:
Yeah, I think this will require tec or me (probably tec :d) playing around with this to see what's wrong. It's probably caused by something how the IPC types are setup in the first place.

Reply:
Can you check in code with the ideal path for this stuff (even though the typechecker will complain, instead of trying a bunch of non-ideal things to work around this issue) so I can take a look?